src/parameterizations/lateral/MOM_streaming_filter.F90

```diff
@@ -0,0 +1,147 @@
+!> Streaming band-pass filter for detecting the instantaneous tidal signals in the simulation
+module MOM_streaming_filter
+
+use MOM_error_handler, only : MOM_mesg, MOM_error, FATAL
+use MOM_hor_index,     only : hor_index_type
+use MOM_time_manager,  only : time_type, time_type_to_real
+use MOM_unit_scaling,  only : unit_scale_type
+
+implicit none ; private
+
+public Filt_register, Filt_accum
+
+#include <MOM_memory.h>
+
+!> The control structure for storing the filter infomation of a particular field
+type, public :: Filter_CS ; private
+  real       :: a, &                   !< Parameter that determines the bandwidth [nondim]
+                om, &                  !< Target frequency of the filter [T-1 ~> s-1]
+                old_time = -1.0        !< The time of the previous accumulating step [T ~> s]
+  real ALLOCABLE_, dimension(NIMEM_,NJMEM_)      :: s1_h, & !< Dummy variable on h grid [A]
```

Comment:
These revisions do resolve the problems with doing math in the halo regions of the arrays, which might cause problems from initialized variables, but there is still a problem here.  The `ALLOCABLE_`, `NIMEM_` and `NJMEM_` cpp macros here are to support either static or dynamic memory for the primary variables that are always used with MOM6.  If you look in `src/framework/MOM_memory_macros.h`, you will see that the use of these macros here means that in static memory mode, all 6 of these arrays will always have memory allocated to  them, even if they are not being used as diagnostics and even though only 2 of the 6 would ever be used.  This can be resolved pretty easily, though, by replacing these macros with their non-static-memory equivalents from `MOM_memory_macros.h` , namely that `ALLOCABLE_` becomes `, allocatable`; `ALLOC()` becomes `allocate()`, and all of the `NIMEM_` pointers become `:`.

  This code would still have some extra copies of the 2-d arrays that could be avoided by duplicating `Filt_accum()` into 3 separate subroutines (e.g., `Filt_accum_h()`, `Filt_accum_u()` and `Filt_accum_v()`) to handle the different staggering of the input variables, and we might want to make this change in a subsequent refactoring step, but if you would prefer I think that we can approve this PR, provided that you make the changes so that this file is always using the non-static-memory version of the code, as described in the paragraph above.

Reply:
Thanks for explaining these details. By using `real, allocatable, dimension(:,:) :: s1`, the memory of `s1` is not pre-allocated, but can be allocated based on the input grid in `Filt_register`. That means I only need to declare one pair of `s1` and `u1`, and keep only one copy of `File_accum()` instead of three of them of different grids. It seems to work when I tested it, but if I miss something here, please let me know.